### ISSUE TITLE
In User Page, "Search for user" filter is slow when facility has a large number of users

### ISSUE THREAD

**Issue by indirectlylit** [learningequality/kolibri#5563]
### Observed behavior

With ~800 users in a facility, the 'Search for a user..." box becomes unusably slow:

> ![image](https://user-images.githubusercontent.com/2367265/58498918-e661e000-8133-11e9-99bf-a8ffd0c3301f.png)

### Expected behavior

The filtering is done client-side, so it should be very snappy.

It can be throttled if necessary, but there's something deeper happening here than just updates happening too often.

### User-facing consequences

frozen browser, wasted battery

### Steps to reproduce

Try filtering users on either the beta or demo server

### Context

0.12.3

**Comment by nucleogenesis** [comment]
@indirectlylit 

The issue seems to be with how many of the components we're rendering and not so much with the sorting/filtering logic.

**Testing Method**

[See this gist.](https://gist.github.com/nucleogenesis/682cd841ddc7c9c7fe484830788ce731) - You can try it by saving it to `file.js` and running `node file.js` wherever you want to locally.

I basically extracted the exact logical flow that is taken to sort the data (with a couple of variables renamed). I also took a dataset of 1030 users for this test.

When I run this with 1030 users in the data, it sorts & filters my requests in  0~75ms. This was what I first tried to see if this is the bottleneck and I'm comfortable saying that the filter/logic code itself is okay.

**Performance Profile**

It seems that the issue is somewhere in the rendering process, particularly with the KDropdownMenu and its children. See the following image profile - performed with ~1000 users.

![search-performance](https://user-images.githubusercontent.com/6356129/58578489-c72f8500-81fd-11e9-9546-f855a8310198.png)

I then removed the entire <KDropdownMenu> section from the table, which notably improved performance but wasn't really all that great. The following is a profile without the Dropdowns.

![search-performance-no-dropdown](https://user-images.githubusercontent.com/6356129/58581823-470d1d80-8205-11e9-824f-2542a7c7bc46.png)

Still - there is an average of 1.5s to load each of the <KLabeledIcons> which are just the left-most columns in the table with the icon + full name - so it's far from snappy. Subjectively speaking - the page is very much faster to load and the filter to respond - but it's not so much better that it would be acceptable in my opinion

**Where to Look Next**

1. Paginate?

If the page didn't have to render so many of each of the components, it would be much faster. Is there a general opposition to pagination for a11y or other design purposes?

Even if the pages were 100 people long, then each page would be pretty quick relatively speaking.

![search-performance-100-users](https://user-images.githubusercontent.com/6356129/58582641-feeefa80-8206-11e9-888e-f0d201e55c17.png)

2. Simplify the Rendering Process?

I suspect this would require a substantial rework of some part of this page or some of the components used on it. I think it would be worth getting on a call to discuss with anybody who might have a good idea of where to start addressing and improving this.

**Comment by indirectlylit** [comment]
It appears that Vue might not be handling the list rendering as efficiently as possible, and is destroying and re-creating the table row DOM nodes (`v-if` behavior) rather than keeping them around and just hiding them (`v-show` behavior).

[The docs claim](https://vuejs.org/v2/guide/list.html#Replacing-an-Array) that shouldn't happen, but perhaps we're doing something with our user table component that confuses its optimizations.

We might try to figure out whether or not vue's "smart" list rendering behavior is being blocked. 

related: https://github.com/learningequality/kolibri/issues/5470

**Comment by nucleogenesis** [comment]
After some more work trying to figure out a way to render everything efficiently, I'm feeling confident in saying that Vue cannot handle rendering this many rows. Note that I'm using ~1000 rows in my testing right now.

1. I'm running on a mid-high end computer and in Chrome's task manager, when I reload the page, I'm seeing CPU > 100% and almost 1G of memory in use while it is rendering before it settles back to ~1% CPU and 500Mb memory. This is just for loading the page in the first place...

2. Here is a comment from an old and long-running issue on Vue's Github https://github.com/vuejs/vue/issues/2000#issuecomment-408512006 where someone profiled and compared React vs Vue for rendering large numbers of plain Jane divs. Our table is far more complex with the nested components and so forth.

tl;dr - 50,000 `<div>hello world</div>`s took 30s in Vue and 5s in React. Not that the React part matters - but the comparison shows that Vue is lacking something in rendering a large number of components.

3. To firmly verify that the sorting/filtering is truly negligible in terms of performance, I added a method to UserTable to slice the users and called that method in the v-for `v-for="user in sliced(users)"`. 

When I slice 10 users, everything is snappy and perfect. When I slice 50 users, I feel a little bit of tolerable lag. At 100, it's pretty bad and it gets almost unusable upwards of that. 

When I'm only rendering 10 users to the screen at once - the text filtering is light speed - the table reacts as I type. At 50+ it hangs before rendering again.

4. I tried moving all of the filtering logic into UserTable. So basically, the UserPage component simply passes the array of users to the UserTable. I moved the sort/filter logic to UserTable. I added a prop so that the `searchFilter` text is passed to the UserTable.

Basically - [the docs](https://vuejs.org/v2/guide/list.html#Replacing-an-Array) @indirectlylit linked to showed the non-mutating methods (ie, filter(), slice()) were working on the component's own data rather than a prop. My thinking was that this may be why Vue re-renders the whole array every time rather than only re-rendering where things changed. So I set the data in UserTable to store the array of users and used a computed property to filter them.

4b. I then adjusted the filtering/sorting logic to just add a `visible: Bool` property on every user and then added a `v-show="user.visible"` thinking that this also might stop Vue from re-rendering the whole thing.

No dice.

5. Even a simpler table with so many rows is slow. In kolibri/plugins/facility_management/assets/src/views/ClassEnrollForm.vue on line 111, I changed the `perPage` to 100 and it's better when you go to `Facility > Classes > Class Details > Enroll Learners` . At 500 it's very laggy and at 1000 it's almost as bad as the UserPage table.

---

Ultimately, even once loaded, my browser feels a tad bit sluggish - I can only imagine what it would be like on a relatively low-powered device.

So, whatever the solution, it seems that it will require that we don't use Vue to render so many rows at once.

A few options that I reviewed are using a virtual-scroller like https://github.com/Aletheios/v-endless-list or https://github.com/Akryum/vue-virtual-scroller - but neither are particularly well-suited for a `<table>`. In an issue on one of them (the more popular of the two) the creator said as much https://github.com/Akryum/vue-virtual-scroller/issues/116 so they would require some hacks using divs and CSS.

Perhaps some homebrewed lazy-loading could work - watching on scroll position and loading more - but this could very well run into the same issue once 100s of rows are scrolled toward and loaded.

It may be worth deciding on and committing to a reusable pagination structure. A quick fix would be to do what is done in kolibri/plugins/facility_management/assets/src/views/ClassEnrollForm.vue - but I know that Devon mentioned his hesitation at approaching more pagination without a more comprehensive approach. FWIW I think this is the most viable solution in any case.

**Comment by indirectlylit** [comment]
> A quick fix would be to do what is done in kolibri/plugins/facility_management/assets/src/views/ClassEnrollForm.vue - but I know that Devon mentioned his hesitation at approaching more pagination without a more comprehensive approach. FWIW I think this is the most viable solution in any case.

Okay, let's go with this strategy for now. I do still believe that our pagination behaviors leave something to be desired, but this will be a straightforward improvement at least.

**Comment by jonboiser** [comment]
#5647 adds a more efficient way to filter larger (~5000 items) lists of FacilityUsers.
